api/src/resources/comment.rs

```diff
@@ -501,13 +501,23 @@ pub struct EntitySpan {
 }
 
 #[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
-pub struct EntityKind(pub String);
+pub struct EntityName(pub String);
 
-impl FromStr for EntityKind {
+#[derive(Debug, Serialize, Deserialize)]
+pub struct EntityDef {
+    name: EntityName,
+    title: String,
+    inherits_from: Vec<String>,
+    trainable: bool,
+}
+
+impl FromStr for EntityDef {
     type Err = Error;
 
     fn from_str(string: &str) -> Result<Self> {
-        Ok(EntityKind(string.to_owned()))
+        serde_json::from_str(string).map_err(|_: SerdeJsonError| Error::BadEntityDefs {
+            entity_defs: string.to_string(),
+        })
```

Comment:
Couple of things here:

- if you're not using the error, you don't need to type it; you can lose the `SerdeJsonError` annotation and corresponding import (I actually don't think you need the type even if you're using it here, Rust should infer it)
- However, I would use the error to show more details to the user!

```suggestion
        serde_json::from_str(string).map_err(|error: SerdeJsonError| Error::BadEntityDef {
            entity_def: string.to_owned(),
            source: error,
        })
```
and see next comment for Error changes

Reply:
I did follow what I could see the convention was for error handling (which I thought was a bit strange to be honest). I see in our errors that we do not resurface the error but instead send back what the user has given as an input.

I do totally agree that surfacing the `json` error does seem like a good idea, so I'll go for your suggestion. :+1:

Reply:
It's a good point. Showing one error inside another is definitively **not** best practice **in a library**.

Ideally, in your library, you're not really concerned with messages that your user would see. At the point it becomes user visible (i.e. in your **binary**), you would handle this case by doing something like:

```rust
let value = parse_my_cli_args().map_err(|error| {
    match error {
        BadEntityDef { entity_def, source } => anyhow!("Bad entity def {}, caused by {}", entity_def, source),
        _ => error,
    }
});
```

However, the point at which we return a result is inside the `structopt` library, so we can't do that here.

You could argue that this whole parsing-from-string is a `cli` application issue, so the code for this should live there. And then convert whatever you parse into the `api` `EntityDef` struct later. But that is very overly complex for this case!